apis/python/src/tiledbsoma/_dataframe.py

```diff
@@ -517,22 +530,39 @@ def _canonicalize_schema(
 def _build_tiledb_schema(
     schema: pa.Schema,
     index_column_names: Sequence[str],
+    domains: Optional[Sequence[Optional[Tuple[Any, Any]]]],
     tiledb_create_options: TileDBCreateOptions,
     context: SOMATileDBContext,
 ) -> tiledb.ArraySchema:
     """Converts an Arrow schema into a TileDB ArraySchema for creation."""
+
+    if domains is None:
+        domains = tuple(None for _ in index_column_names)
+    else:
+        ndom = len(domains)
+        nidx = len(index_column_names)
+        if ndom != nidx:
+            raise ValueError(
+                f"if domains are specified, they must have the same length as index_column_names; got {ndom} != {nidx}"
+            )
+
     dims = []
-    for index_column_name in index_column_names:
+    for index_column_name, domain in zip(index_column_names, domains):
         pa_type = schema.field(index_column_name).type
         dtype = _arrow_types.tiledb_type_from_arrow_type(
             pa_type, is_indexed_column=True
         )
-        domain: Tuple[Any, Any]
-        if isinstance(dtype, str):
+
+        if domain is not None:
+            pass  # User-specified; go with it
+
+        elif isinstance(dtype, str):
             domain = None, None
         elif np.issubdtype(dtype, NPInteger):
             iinfo = np.iinfo(cast(NPInteger, dtype))
             domain = iinfo.min, iinfo.max - 1
+            if index_column_name == SOMA_JOINID:
+                domain = (0, domain[1])
```

Comment:
I'm not sure this approach is consistent with how we handle schema checks for soma_joinid.

In the schema, if the user specifies the type, it must be correct or we throw an error.  We don't silently correct it.

In this case, if the domain is wrong, the code silently corrects the domain.  I don't think there is a strong case one way or the other, but perhaps we should be consistent.  I.e., check that a user-specified domain is correct, and then simply raise an error if it is not.

optional

Reply:
@bkmartinjr this isn't correcting a user error.

This block only applies when the user has _not_ specified the domain. Here we _choose_ one, and we specifically disallow negative `soma_joinid` which is in conformance with the spec.

https://github.com/single-cell-data/SOMA/pull/124
https://github.com/single-cell-data/SOMA/issues/121

Reply:
You are absolutely correct - apologies.

That does raise a possible follow-up: do we error check that the domain for `soma_joinid` is bounded by [0, maxint)?

Reply:
@bkmartinjr good call & i'll resolve this comment in favor of that one